backend/app/notify/email.go

```diff
@@ -0,0 +1,380 @@
+package notify
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"net"
+	"net/smtp"
+	"strings"
+	"sync"
+	"text/template"
+	"time"
+
+	log "github.com/go-pkgz/lgr"
+	"github.com/go-pkgz/repeater"
+	"github.com/pkg/errors"
+)
+
+// EmailParams contain settings for email set up
+type EmailParams struct {
+	Host                 string        // SMTP host
+	Port                 int           // SMTP port
+	TLS                  bool          // TLS auth
+	From                 string        // From email field
+	Username             string        // user name
+	Password             string        // password
+	TimeOut              time.Duration // TLS connection timeout
+	MsgTemplate          string        // request message template
+	VerificationTemplate string        // verification message template
+	BufferSize           int           // email send buffer size
+	FlushDuration        time.Duration // maximum time after which email will me sent, 30s by default
+}
+
+// Email implements notify.Destination for email
+type Email struct {
+	EmailParams
+	smtpClient // embedded into Email for test purposes: not set in production code, used in autoFlush initialisation
+
+	msgTmpl    *template.Template // parsed request message template
+	verifyTmpl *template.Template // parsed verification message template
+	submit     chan emailMessage  // unbuffered channel for email sending
+	once       sync.Once
+}
+
+// smtpClient interface defines subset of net/smtp used by email client
+type smtpClient interface {
+	Mail(string) error
+	Auth(smtp.Auth) error
+	Rcpt(string) error
+	Data() (io.WriteCloser, error)
+	Quit() error
+	Close() error
+}
+
+type emailMessage struct {
+	message string
+	to      string
+}
+
+// msgTmplData store data for message from request template execution
+type msgTmplData struct {
+	From      string
+	To        string
+	Orig      string
+	Link      string
+	PostTitle string
+}
+
+// verifyTmplData store data for verification message template execution
+type verifyTmplData struct {
+	User  string
+	Email string
+	Token string
+	Site  string
+}
+
+const defaultEmailTimeout = 10 * time.Second
+const defaultFlushDuration = time.Second * 30
+const defaultEmailTemplate = `{{.From}}{{if .To}} → {{.To}}{{end}}
+
+{{.Orig}}
+
+↦ <a href="{{.Link}}">{{if .PostTitle}}{{.PostTitle}}{{else}}original comment{{end}}</a>
+`
+
+const defaultEmailVerificationTemplate = `Confirmation for {{.User}} {{.Email}}, site {{.Site}}
+
+Token: {{.Token}}
+`
+
+//NewEmail makes new Email object, returns it even in case of problems
+// (e.MsgTemplate parsing error or error while testing smtp connection by credentials provided in params)
+func NewEmail(params EmailParams) (*Email, error) {
```

Comment:
I'll look into switching NewEmail to Options (however I see it would make code more complex and no benefits in it taking in consideration we have single call of this function at the moment).

Reply:
Pleases see [initialisation example](https://github.com/paskal/remark/blob/email_notifications/backend/app/cmd/server.go#L342-L358): we need to pass everything from `cmd.Server` and I wonder if functional options are really necessary here.

Reply:
https://github.com/paskal/remark/pull/3 is an illustration of how this new Email will be used.